tests/tamper_robustness.py

```diff
@@ -0,0 +1,274 @@
+
+    with primary.client("user0") as c:
+        node_root = get_node_root_dir(primary)
+
+        desired_non_empty_dirs = [get_index_dir(node_root), get_ledger_dir(node_root)]
+
+        def empty_dir(p):
+            return len(os.listdir(p)) == 0
+
+        loops = 0
+        loop_limit = 10
+        while any(empty_dir(p) for p in desired_non_empty_dirs):
+            if loops > loop_limit:
+                raise RuntimeError(
+                    f"Failed to produce ledger and index after {loop_limit} loops"
+                )
+            loops += 1
+
+            # Post several new entries, then fetch the latest historically
+            num_entries = 500
+            LOG.info(f"Posting {num_entries} new entries in loop #{loops}")
+            for i in range(num_entries):
+                r = c.post(
+                    "/app/log/public",
+                    {"id": TARGET_ID, "msg": f"Hello world: {loops}:{i}"},
+                    log_capture=[],
+                )
+                assert r.status_code == http.HTTPStatus.OK, (
+                    r.status_code,
+                    r.body.text(),
+                )
+
+            infra.commit.wait_for_commit(c, seqno=r.seqno, view=r.view)
+
+            get_all_entries(c, TARGET_ID, from_seqno=r.seqno, to_seqno=r.seqno)
+
+        LOG.info("Populated interesting dirs:")
+        for p in desired_non_empty_dirs:
+            LOG.info(f"  {p} contains {len(os.listdir(p))} files")
+
+
+@contextlib.contextmanager
+def hide_file(src_path):
+    with tempfile.TemporaryDirectory(dir=os.path.dirname(src_path)) as tmp_dir_name:
+        dst_path = os.path.join(tmp_dir_name, os.path.basename(src_path))
+        LOG.warning(f"Temporarily hiding file {src_path} (backed up at {dst_path})")
+        shutil.move(src_path, dst_path)
+        yield
+        shutil.move(dst_path, src_path)
+
+
+@contextlib.contextmanager
+def truncate_file(src_path):
+    with tempfile.TemporaryDirectory(dir=os.path.dirname(src_path)) as tmp_dir_name:
+        dst_path = os.path.join(tmp_dir_name, os.path.basename(src_path))
+        LOG.warning(f"Temporarily truncating file {src_path} (backed up at {dst_path})")
+        shutil.copy(src_path, dst_path)
+        src_len = os.path.getsize(src_path)
+        with open(src_path, "rb+") as f:
+            f.truncate(random.randrange(src_len // 2, src_len - 1))
+        yield
+        shutil.move(dst_path, src_path)
+
+
+@contextlib.contextmanager
+def modify_file(src_path):
+    with tempfile.TemporaryDirectory(dir=os.path.dirname(src_path)) as tmp_dir_name:
+        dst_path = os.path.join(tmp_dir_name, os.path.basename(src_path))
+        LOG.warning(f"Temporarily modifying file {src_path} (backed up at {dst_path})")
+        shutil.copy(src_path, dst_path)
+        with open(src_path, "rb+") as f:
+            content = list(f.read())
+            size = len(content)
+            for idx in (size // 3, size // 2, int(size // 1.5)):
+                content[idx] //= 2
```

Comment:
0 stays 0? If the files are big enough it wouldn't make much of a difference I guess. Maybe `content[idx] ^ 0b01010101`?